File: server/src/services/custom_quote_service.ts
import { PrismaClient } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';
import { CustomQuoteInput, PageResult } from '@/types';
import memberService from './member_service';

const prisma = new PrismaClient();

const FREE_QUOTA = 50;

async function list(
    userId: number,
    page: number,
    pageSize: number
): Promise<PageResult<CustomQuoteItem>> {
    const skip = (page - 1) * pageSize;

    const [total, records] = await Promise.all([
        prisma.customQuote.count({ where: { userId } }),
        prisma.customQuote.findMany({
            where: { userId },
            skip,
            take: pageSize,
            orderBy: { createdAt: 'desc' },
        }),
    ]);

    const list: CustomQuoteItem[] = records.map((r) => ({
        id: String(r.id),
        content: r.content,
        author: r.author ?? undefined,
        source: r.source ?? undefined,
        note: r.note ?? undefined,
        tags: r.tags ?? undefined,
        createdAt: r.createdAt.toISOString(),
    }));

    return { total, list };
}

interface CustomQuoteItem {
    id: string;
    content: string;
    author?: string;
    source?: string;
    note?: string;
    tags?: string;
    createdAt: string;
}

async function create(userId: number, input: CustomQuoteInput): Promise<CustomQuoteItem> {
    const quota = await getQuota(userId);
    if (quota.remaining <= 0) {
        throw new BusinessError(403, '自定义名句数量已达上限，请开通会员');
    }

    const record = await prisma.customQuote.create({
        data: {
            userId,
            content: input.content,
            author: input.author,
            source: input.sourceName,
            note: input.note,
            tags: input.categoryIds?.join(','),
        },
    });

    return {
        id: String(record.id),
        content: record.content,
        author: record.author ?? undefined,
        source: record.source ?? undefined,
        note: record.note ?? undefined,
        tags: record.tags ?? undefined,
        createdAt: record.createdAt.toISOString(),
    };
}

async function update(
    userId: number,
    id: number,
    input: CustomQuoteInput
): Promise<CustomQuoteItem> {
    const existing = await prisma.customQuote.findUnique({ where: { id } });
    if (!existing) throw new BusinessError(404, '自定义名句不存在');
    if (existing.userId !== userId) throw new BusinessError(403, '无权操作');

    const updated = await prisma.customQuote.update({
        where: { id },
        data: {
            content: input.content,
            author: input.author,
            source: input.sourceName,
            note: input.note,
            tags: input.categoryIds?.join(','),
        },
    });

    return {
        id: String(updated.id),
        content: updated.content,
        author: updated.author ?? undefined,
        source: updated.source ?? undefined,
        note: updated.note ?? undefined,
        tags: updated.tags ?? undefined,
        createdAt: updated.createdAt.toISOString(),
    };
}

async function remove(userId: number, id: number): Promise<void> {
    const existing = await prisma.customQuote.findUnique({ where: { id } });
    if (!existing) throw new BusinessError(404, '自定义名句不存在');
    if (existing.userId !== userId) throw new BusinessError(403, '无权操作');

    await prisma.customQuote.delete({ where: { id } });
}

async function getQuota(userId: number): Promise<{ total: number; used: number; remaining: number }> {
    const isMember = await memberService.checkIsMember(userId);

    const used = await prisma.customQuote.count({ where: { userId } });
    const total = isMember ? Infinity : FREE_QUOTA;
    const remaining = isMember ? Infinity : Math.max(0, FREE_QUOTA - used);

    return { total, used, remaining };
}

export { list, create, update, remove, getQuota };

File: server-py/app/services/custom_quote_service.py
"""自定义名句服务（对应 TS custom_quote_service.ts）"""
from datetime import datetime

from sqlalchemy.orm import Session

from ..deps import BusinessError
from ..models import CustomQuote
from .member_service import checkIsMember

FREE_QUOTA = 50


def _toItem(r: CustomQuote) -> dict:
    return {
        "id": str(r.id),
        "content": r.content,
        "author": r.author,
        "source": r.source,
        "note": r.note,
        "tags": r.tags,
        "createdAt": r.createdAt.isoformat() if r.createdAt else None,
    }


def _joinTags(categoryIds) -> str | None:
    if categoryIds is None:
        return None
    return ",".join(str(x) for x in categoryIds)


def list(db: Session, userId: int, page: int, pageSize: int) -> dict:
    skip = (page - 1) * pageSize

    total = db.query(CustomQuote).filter(CustomQuote.userId == userId).count()
    records = (
        db.query(CustomQuote)
        .filter(CustomQuote.userId == userId)
        .order_by(CustomQuote.createdAt.desc())
        .offset(skip)
        .limit(pageSize)
        .all()
    )

    return {"total": total, "list": [_toItem(r) for r in records]}


def create(db: Session, userId: int, input: dict) -> dict:
    quota = getQuota(db, userId)
    if quota["remaining"] <= 0:
        raise BusinessError(403, "自定义名句数量已达上限，请开通会员")

    record = CustomQuote(
        userId=userId,
        content=input.get("content"),
        author=input.get("author"),
        source=input.get("sourceName"),
        note=input.get("note"),
        tags=_joinTags(input.get("categoryIds")),
    )
    db.add(record)
    db.commit()
    db.refresh(record)

    return _toItem(record)


def update(db: Session, userId: int, id: int, input: dict) -> dict:
    existing = db.query(CustomQuote).filter(CustomQuote.id == id).first()
    if not existing:
        raise BusinessError(404, "自定义名句不存在")
    if existing.userId != userId:
        raise BusinessError(403, "无权操作")

    existing.content = input.get("content")
    existing.author = input.get("author")
    existing.source = input.get("sourceName")
    existing.note = input.get("note")
    existing.tags = _joinTags(input.get("categoryIds"))
    existing.updatedAt = datetime.now()
    db.commit()
    db.refresh(existing)

    return _toItem(existing)


def remove(db: Session, userId: int, id: int) -> None:
    existing = db.query(CustomQuote).filter(CustomQuote.id == id).first()
    if not existing:
        raise BusinessError(404, "自定义名句不存在")
    if existing.userId != userId:
        raise BusinessError(403, "无权操作")

    db.delete(existing)
    db.commit()


def getQuota(db: Session, userId: int) -> dict:
    isMember = checkIsMember(db, userId)

    used = db.query(CustomQuote).filter(CustomQuote.userId == userId).count()
    total = float("inf") if isMember else FREE_QUOTA
    remaining = float("inf") if isMember else max(0, FREE_QUOTA - used)

    return {"total": total, "used": used, "remaining": remaining}
